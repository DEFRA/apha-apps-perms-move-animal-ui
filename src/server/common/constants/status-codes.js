/**
 * @typedef {Record<string, number>} StatusCodes
 */
export const statusCodes = {
  ok: 200,
  noContent: 204,
  badRequest: 400,
  unauthorized: 401,
  forbidden: 403,
  notFound: 404,
  imATeapot: 418,
  redirect: 302,
  serverError: 500,
  badGateway: 502,
  gatewayTimeout: 504
}
