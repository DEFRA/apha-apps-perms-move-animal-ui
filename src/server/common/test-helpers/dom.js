import { JSDOM } from 'jsdom'

/**
 * @param {string} payload
 * @returns {Document}
 */
export const parseDocument = (payload) => new JSDOM(payload).window.document
