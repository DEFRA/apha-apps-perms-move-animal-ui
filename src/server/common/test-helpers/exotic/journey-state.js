export const aboutSectionVisitComplete = {
  movementType: 'visit'
}

export const aboutSectionNonVisitComplete = {
  movementType: 'onto-premises',
  whatIsMoving: 'carcasses',
  whatAreYouMovingDetails: 'pig carcasses',
  howMuchAreYouMoving: '10 pig carcasses',
  whatAreYouMovingDescription: 'culled pig carcasses for incineration'
}
