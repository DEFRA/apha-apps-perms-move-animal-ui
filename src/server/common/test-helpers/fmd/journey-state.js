export const aboutSectionComplete = {
  movementActivityType: 'slaughter-onsite',
  animalSlaughtered: 'cattle',
  slaughteredNumber: 123,
  animalIds: 'ear tags'
}
