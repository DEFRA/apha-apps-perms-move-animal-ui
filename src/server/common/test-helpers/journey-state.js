/** @import { OnOffFarmData } from '~/src/server/common/model/answer/on-off-farm/on-off-farm.js' */
/** @import { DestinationTypeData } from '~/src/server/common/model/answer/destination-type/destination-type.js' */
/** @import { ReasonForMovementData } from '~/src/server/common/model/answer/reason-for-movement/reason-for-movement.js' */
/** @import { QuantityOptionsData } from '~/src/server/common/model/answer/quantity-options/quantity-options.js' */
/** @import { QuantityHalfHerdData } from '~/src/server/common/model/answer/quantity-half-herd/quantity-half-herd.js' */

const validCphNumber = '12/345/6789'
const validOriginType = 'afu'
/** @type {OnOffFarmData} */
const validOnOffFarm = 'off'
const validAddress = {
  addressLine1: 'Starfleet Headquarters',
  addressTown: 'San Francisco',
  addressPostcode: 'RG24 8RR'
}

export const validOriginSectionState = {
  onOffFarm: validOnOffFarm,
  originType: validOriginType,
  cphNumber: validCphNumber,
  address: validAddress
}

/** @type {DestinationTypeData} */
const testDestinationType = 'slaughter'

export const validDestinationSectionState = {
  destinationType: testDestinationType,
  destinationFarmCph: validCphNumber,
  destinationFarmAddress: validAddress,
  maxNumberOfAnimals: '1',
  /** @type {ReasonForMovementData} */
  reasonForMovement: 'routineRestocking',
  /** @type {QuantityOptionsData} */
  movingMoreThan75Animals: 'no',
  /** @type {QuantityHalfHerdData} */
  movingMoreThanHalfExistingHerd: 'yes'
}

export const validIdentificationSectionState = {
  earTags: 'some-ear-tags'
}

export const validBiosecuritySectionState = {
  keptSeparately: 'yes',
  grazing: 'yes',
  lastGrazed: 'yesterday',
  manureAndSlurry: 'yes',
  grazingFieldHowSeparated: 'some details',
  roadsAndTracks: 'yes',
  buildingsAnyShared: 'yes',
  buildingsHowMinimiseContamination: 'somehow',
  peopleDisinfection: 'ppe',
  disinfectant: 'some disinfectant',
  dilutionRate: '15',
  badgers: 'some text'
}

export const validBiosecurityMapSectionState = {
  'upload-plan': {
    metadata: {
      uploadId: '41572cf8-2e37-495e-9ad2-0b0f23f1b277',
      uploadUrl:
        'http://localhost:7337/upload-and-scan/41572cf8-2e37-495e-9ad2-0b0f23f1b277',
      statusUrl:
        'http://localhost:7337/status/41572cf8-2e37-495e-9ad2-0b0f23f1b277'
    },
    status: {
      uploadStatus: 'ready',
      metadata: {},
      form: {
        crumb: 'QVJdAVFWpx90BqITFf6tFf7CpwJFNn2jGN-8CyKwlO9',
        nextPage: '',
        file: {
          fileId: '3d3c2a09-2888-4199-9bd6-ac7eda3125f0',
          filename: '34998B77-FB3E-44DB-BC0E-05154D6549E0.jpeg',
          contentType: 'image/jpeg',
          fileStatus: 'complete',
          contentLength: 374478,
          checksumSha256: '3etoXNlR16WpgCiwylqccFxLVg3OrZvpGUqmigmrhcU=',
          detectedContentType: 'image/jpeg',
          s3Key:
            'biosecurity-map/41572cf8-2e37-495e-9ad2-0b0f23f1b277/3d3c2a09-2888-4199-9bd6-ac7eda3125f0',
          s3Bucket: 'apha'
        }
      },
      numberOfRejectedFiles: 0
    }
  }
}

export const validLicenceSectionState = {
  fullName: {
    firstName: 'Kathryn',
    lastName: 'Janeway'
  },
  receiveMethod: 'email',
  emailAddress: 'kathryn@starfleet.com'
}

export const validApplicationStateWithBioSecurity = {
  origin: {
    ...validOriginSectionState,
    onOffFarm: 'on',
    originType: 'market'
  },
  destination: {
    ...validDestinationSectionState,
    destinationType: 'tb-restricted-farm'
  },
  identification: validIdentificationSectionState,
  licence: validLicenceSectionState,
  biosecurity: validBiosecuritySectionState,
  'biosecurity-map': validBiosecurityMapSectionState
}
