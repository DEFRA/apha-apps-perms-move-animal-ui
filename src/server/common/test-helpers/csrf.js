import { statusCodes } from '~/src/server/common/constants/status-codes.js'
/* global expect, test */

/**
 * @param {OutgoingHttpHeaders} headers
 * @param {string} name
 * @returns {string}
 */
const findCookie = (headers, name) => {
  const setCookie = headers['set-cookie']

  /** @type {string[] | undefined} */
  const cookies = typeof setCookie === 'string' ? [setCookie] : setCookie
  return cookies?.find((h) => h?.includes(`${name}=`)) ?? ''
}

/**
 * @param {string} cookieString
 */
const extractCookieValue = (cookieString) =>
  cookieString.split('=')[1].split(';')[0]

/**
 * @param {function(): Server} serverFn
 * @param {ServerInjectOptions} injectionOptions
 */
export const testCsrfProtectedPost = (serverFn, injectionOptions) =>
  test('should reject if experiencing an apparent csrf attack (they have their session cookie, but no hidden form value)', async () => {
    const { statusCode } = await serverFn().inject({
      ...injectionOptions,
      headers: {
        ...(injectionOptions.headers ?? {}),
        Cookie: 'crumb=csrf-value'
      }
    })

    expect(statusCode).toEqual(statusCodes.forbidden)
  })

/**
 * @param {function(): Server} serverFn
 * @param {ServerInjectOptions} injectionOptions
 */
export const testCsrfProtectedGet = (serverFn, injectionOptions) =>
  test('should contain a cookie with a set CSRF value, and contain a hidden form field to submit that CSRF back to the server', async () => {
    const { headers, payload } = await serverFn().inject(injectionOptions)

    const token = extractCookieValue(findCookie(headers, 'crumb'))
    expect(token).not.toHaveLength(0)
    expect(payload).toContain(
      `<input type="hidden" name="crumb" value="${token}"`
    )
  })

/**
 * @param {ServerInjectOptions} injectionOptions
 * @returns {ServerInjectOptions}
 */
export const withCsrfProtection = (injectionOptions) => ({
  ...injectionOptions,
  payload: {
    ...(typeof injectionOptions.payload === 'object'
      ? injectionOptions.payload
      : {}),
    crumb: 'csrf-value'
  },
  headers: {
    ...(injectionOptions.headers ?? {}),
    Cookie: 'crumb=csrf-value'
  }
})

/**
 * @import { OutgoingHttpHeaders } from 'http'
 * @import { ServerInjectOptions, Server } from '@hapi/hapi'
 */
