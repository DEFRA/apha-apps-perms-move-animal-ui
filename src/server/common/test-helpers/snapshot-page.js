import { parseDocument } from '../../common/test-helpers/dom.js'
import { createServer } from '~/src/server/index.js'
import { withCsrfProtection } from './csrf.js'
/** @import { Server } from '@hapi/hapi' */

/* global expect, it, beforeAll, afterAll, describe */

/**
 * @param {string} describeMessage
 * @param {string} shouldMessage
 * @param {string} urlPath
 */
export const describePageSnapshot = (
  describeMessage,
  shouldMessage,
  urlPath
) => {
  describe(describeMessage, () => {
    /** @type {Server} */
    let server

    beforeAll(async () => {
      server = await createServer()
      await server.initialize()
    })

    afterAll(async () => {
      await server.stop({ timeout: 0 })
    })

    it(shouldMessage, async () => {
      const { payload } = await server.inject(
        withCsrfProtection({
          method: 'GET',
          url: urlPath
        })
      )

      const content =
        parseDocument(payload).querySelector('#main-content')?.innerHTML

      expect(content).toMatchSnapshot()
    })
  })
}
