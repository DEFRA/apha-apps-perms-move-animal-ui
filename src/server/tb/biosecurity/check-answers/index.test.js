import { createServer } from '~/src/server/index.js'
import { statusCodes } from '~/src/server/common/constants/status-codes.js'
import { withCsrfProtection } from '~/src/server/common/test-helpers/csrf.js'
import { parseDocument } from '~/src/server/common/test-helpers/dom.js'
import SessionTestHelper from '../../../common/test-helpers/session-helper.js'
import { biosecuritySummaryPage } from './index.js'
import { describePageSnapshot } from '../../../common/test-helpers/snapshot-page.js'
import { keptSeparatelyPage } from '../kept-separately/index.js'
import { validBiosecuritySectionState } from '../../../common/test-helpers/journey-state.js'

const pageUrl = '/biosecurity/check-answers'

describe('#biosecuritySummaryPage', () => {
  /** @type {Server} */
  let server
  /** @type {SessionTestHelper} */
  let session

  beforeAll(async () => {
    server = await createServer()
    await server.initialize()
    session = await SessionTestHelper.create(server)
  })

  afterAll(async () => {
    await server.stop({ timeout: 0 })
  })

  it('should render expected response when default state present', async () => {
    await session.setSectionState('biosecurity', validBiosecuritySectionState)

    const { payload, statusCode } = await server.inject(
      withCsrfProtection(
        {
          method: 'GET',
          url: pageUrl
        },
        {
          Cookie: session.sessionID
        }
      )
    )

    const document = parseDocument(payload)
    expect(statusCode).toBe(statusCodes.ok)
    expect(document.title).toBe(biosecuritySummaryPage.pageTitle)

    expect(payload).toEqual(expect.stringContaining('Yes'))
  })

  it('should redirect user to kept separately page if they`ve not selected a value', async () => {
    await session.setSectionState('biosecurity', {
      ...validBiosecuritySectionState,
      keptSeparately: null
    })

    const { statusCode, headers } = await server.inject(
      withCsrfProtection(
        {
          method: 'GET',
          url: pageUrl
        },
        {
          Cookie: session.sessionID
        }
      )
    )

    expect(statusCode).toBe(statusCodes.redirect)
    expect(statusCode).toBe(statusCodes.redirect)
    expect(headers.location).toBe(
      `${keptSeparatelyPage.urlPath}?redirect_uri=${biosecuritySummaryPage.urlPath}`
    )
  })
})

describePageSnapshot({
  describes: '#biosecuritySummaryPage.content',
  it: 'should render the expected content',
  pageUrl,
  state: {
    application: { biosecurity: validBiosecuritySectionState }
  }
})

/**
 * @import { Server } from '@hapi/hapi'
 */
