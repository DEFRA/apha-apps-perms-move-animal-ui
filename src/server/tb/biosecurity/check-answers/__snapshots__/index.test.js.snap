// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#biosecuritySummaryPage.content should render the expected content 1`] = `
"
  <div class="govuk-grid-row">
    <div class="govuk-grid-column-two-thirds">
      <h1 class="govuk-heading-l">Check your answers before you continue your application</h1>
    </div>
  </div>
    <form method="POST" novalidate="">
      <div class="govuk-grid-row">
        <div class="govuk-grid-column-two-thirds">
          <input type="hidden" name="crumb" value="csrf-value">
          <input type="hidden" name="nextPage" value="">
  
  <dl class="govuk-summary-list govuk-!-margin-bottom-9">
  <div class="govuk-summary-list__row">
    <dt class="govuk-summary-list__key govuk-!-width-one-half govuk-!-font-weight-regular">
      Will you separate the incoming animals from the resident herd?
    </dt>
    <dd class="govuk-summary-list__value">
      Yes
    </dd>
    <dd class="govuk-summary-list__actions">
      <a class="govuk-link" href="/biosecurity/kept-separately?redirect_uri=/biosecurity/check-answers" data-testid="keptSeparately-change-link">Change<span class="govuk-visually-hidden"> Will you separate the incoming animals from the resident herd?</span></a>
    </dd>
  </div>
  <div class="govuk-summary-list__row">
    <dt class="govuk-summary-list__key govuk-!-width-one-half govuk-!-font-weight-regular">
      Will the incoming animals be grazed?
    </dt>
    <dd class="govuk-summary-list__value">
      Yes
    </dd>
    <dd class="govuk-summary-list__actions">
      <a class="govuk-link" href="/biosecurity/grazing?redirect_uri=/biosecurity/check-answers" data-testid="grazing-change-link">Change<span class="govuk-visually-hidden"> Will the incoming animals be grazed?</span></a>
    </dd>
  </div>
  <div class="govuk-summary-list__row">
    <dt class="govuk-summary-list__key govuk-!-width-one-half govuk-!-font-weight-regular">
      How long ago was the field last grazed by the resident herd?
    </dt>
    <dd class="govuk-summary-list__value">
      yesterday
    </dd>
    <dd class="govuk-summary-list__actions">
      <a class="govuk-link" href="/biosecurity/last-grazed?redirect_uri=/biosecurity/check-answers" data-testid="lastGrazed-change-link">Change<span class="govuk-visually-hidden"> How long ago was the field last grazed by the resident herd?</span></a>
    </dd>
  </div>
  <div class="govuk-summary-list__row">
    <dt class="govuk-summary-list__key govuk-!-width-one-half govuk-!-font-weight-regular">
      Has any manure or slurry been put on the grazing field in the past 60 days?
    </dt>
    <dd class="govuk-summary-list__value">
      Yes
    </dd>
    <dd class="govuk-summary-list__actions">
      <a class="govuk-link" href="/biosecurity/manure-and-slurry?redirect_uri=/biosecurity/check-answers" data-testid="manureAndSlurry-change-link">Change<span class="govuk-visually-hidden"> Has any manure or slurry been put on the grazing field in the past 60 days?</span></a>
    </dd>
  </div>
  <div class="govuk-summary-list__row">
    <dt class="govuk-summary-list__key govuk-!-width-one-half govuk-!-font-weight-regular">
      Which measures are being taken to reduce the spread of TB when the animals are grazing?
    </dt>
    <dd class="govuk-summary-list__value">
      A minimum of 3 metres will separate the animals
    </dd>
    <dd class="govuk-summary-list__actions">
      <a class="govuk-link" href="/biosecurity/grazing-field-how-separated?redirect_uri=/biosecurity/check-answers" data-testid="grazingFieldHowSeparated-change-link">Change<span class="govuk-visually-hidden"> Which measures are being taken to reduce the spread of TB when the animals are grazing?</span></a>
    </dd>
  </div>
  <div class="govuk-summary-list__row">
    <dt class="govuk-summary-list__key govuk-!-width-one-half govuk-!-font-weight-regular">
      How will you manage slurry and manure?
    </dt>
    <dd class="govuk-summary-list__value">
      Manure and slurry is securely stored so domestic and wild animals can't access it
    </dd>
    <dd class="govuk-summary-list__actions">
      <a class="govuk-link" href="/biosecurity/manure-and-slurry-details?redirect_uri=/biosecurity/check-answers" data-testid="manureAndSlurryDetails-change-link">Change<span class="govuk-visually-hidden"> How will you manage slurry and manure?</span></a>
    </dd>
  </div>
  <div class="govuk-summary-list__row">
    <dt class="govuk-summary-list__key govuk-!-width-one-half govuk-!-font-weight-regular">
      What disinfectant are you using?
    </dt>
    <dd class="govuk-summary-list__value">
      some disinfectant
    </dd>
    <dd class="govuk-summary-list__actions">
      <a class="govuk-link" href="/biosecurity/disinfectant?redirect_uri=/biosecurity/check-answers" data-testid="disinfectant-change-link">Change<span class="govuk-visually-hidden"> What disinfectant are you using?</span></a>
    </dd>
  </div>
  <div class="govuk-summary-list__row">
    <dt class="govuk-summary-list__key govuk-!-width-one-half govuk-!-font-weight-regular">
      Confirmation of the dilution rate
    </dt>
    <dd class="govuk-summary-list__value">
      I confirm a dilution rate of 1:15 is used on the farm or premises
    </dd>
    <dd class="govuk-summary-list__actions">
      <a class="govuk-link" href="/biosecurity/disinfectant-dilution?redirect_uri=/biosecurity/check-answers" data-testid="dilutionRate-change-link">Change<span class="govuk-visually-hidden"> Confirmation of the dilution rate</span></a>
    </dd>
  </div>
  <div class="govuk-summary-list__row">
    <dt class="govuk-summary-list__key govuk-!-width-one-half govuk-!-font-weight-regular">
      Will the incoming animals share any equipment and machinery with the resident herd?
    </dt>
    <dd class="govuk-summary-list__value">
      Yes
    </dd>
    <dd class="govuk-summary-list__actions">
      <a class="govuk-link" href="/biosecurity/equipment-any-shared?redirect_uri=/biosecurity/check-answers" data-testid="equipmentShared-change-link">Change<span class="govuk-visually-hidden"> Will the incoming animals share any equipment and machinery with the resident herd?</span></a>
    </dd>
  </div>
  <div class="govuk-summary-list__row">
    <dt class="govuk-summary-list__key govuk-!-width-one-half govuk-!-font-weight-regular">
      Which measures are in place to clean and disinfect equipment to reduce the risk of spreading TB?
    </dt>
    <dd class="govuk-summary-list__value">
      Designated disinfection points
    </dd>
    <dd class="govuk-summary-list__actions">
      <a class="govuk-link" href="/biosecurity/equipment-how-minimise-contamination?redirect_uri=/biosecurity/check-answers" data-testid="equipmentHowMinimiseContamination-change-link">Change<span class="govuk-visually-hidden"> Which measures are in place to clean and disinfect equipment to reduce the risk of spreading TB?</span></a>
    </dd>
  </div>
  <div class="govuk-summary-list__row">
    <dt class="govuk-summary-list__key govuk-!-width-one-half govuk-!-font-weight-regular">
      Will the incoming animals be housed?
    </dt>
    <dd class="govuk-summary-list__value">
      Yes
    </dd>
    <dd class="govuk-summary-list__actions">
      <a class="govuk-link" href="/biosecurity/buildings-any-shared?redirect_uri=/biosecurity/check-answers" data-testid="animalsHoused-change-link">Change<span class="govuk-visually-hidden"> Will the incoming animals be housed?</span></a>
    </dd>
  </div>
  <div class="govuk-summary-list__row">
    <dt class="govuk-summary-list__key govuk-!-width-one-half govuk-!-font-weight-regular">
      Which measures are being taken to reduce the spread of TB during housing?
    </dt>
    <dd class="govuk-summary-list__value">
      Isolation of incoming animals in a separate building, shed or pen on arrival
    </dd>
    <dd class="govuk-summary-list__actions">
      <a class="govuk-link" href="/biosecurity/buildings-how-minimise-contamination?redirect_uri=/biosecurity/check-answers" data-testid="buildingsHowMinimiseContamination-change-link">Change<span class="govuk-visually-hidden"> Which measures are being taken to reduce the spread of TB during housing?</span></a>
    </dd>
  </div>
  <div class="govuk-summary-list__row">
    <dt class="govuk-summary-list__key govuk-!-width-one-half govuk-!-font-weight-regular">
      Which measures are staff taking to reduce the risk of spreading TB?
    </dt>
    <dd class="govuk-summary-list__value">
      Dedicated clothing and personal protective equipment (PPE)
    </dd>
    <dd class="govuk-summary-list__actions">
      <a class="govuk-link" href="/biosecurity/people-disinfection?redirect_uri=/biosecurity/check-answers" data-testid="peopleDisinfection-change-link">Change<span class="govuk-visually-hidden"> Which measures are staff taking to reduce the risk of spreading TB?</span></a>
    </dd>
  </div>
  <div class="govuk-summary-list__row">
    <dt class="govuk-summary-list__key govuk-!-width-one-half govuk-!-font-weight-regular">
      Which measures are you taking to reduce the risk of infection from wildlife?
    </dt>
    <dd class="govuk-summary-list__value">
      Badger-proof fencing, such as wired fences, solid aluminium sheeted gates and rail fences or retractable electric fences
    </dd>
    <dd class="govuk-summary-list__actions">
      <a class="govuk-link" href="/biosecurity/badgers?redirect_uri=/biosecurity/check-answers" data-testid="badgers-change-link">Change<span class="govuk-visually-hidden"> Which measures are you taking to reduce the risk of infection from wildlife?</span></a>
    </dd>
  </div>
</dl>

            <button type="submit" class="govuk-button" data-module="govuk-button" id="continue-button">
  Continue
</button>
        </div>
      </div>
    </form>
        "
`;
